File: BaekJoon/Silver/9012/Python/9012.py
import sys

def isVPS(string):
    stack = []
    for char in string:
        if char == '(':
            stack.append('(')
        else:
            if len(stack) == 0:
                return False
            elif stack.pop() != '(':
                return False
    if len(stack) != 0:
        return False
    else:
        return True

t = int(sys.stdin.readline())
for i in range(0, t):
    str1 = sys.stdin.readline().rstrip()
    print('YES' if isVPS(str1) else 'NO')

File: BaekJoon/Silver/9012/C++/9012.cpp
// no.9012: 괄호 (S4)

#include <cstdio>
#include <stack>
#include <cstring>
using namespace std;

bool IsValidPS(const char* str);

int main() {
    int n;
    scanf("%d", &n);
    getchar();

    char str[52];
    for(int i=0; i<n; i++) {
        fgets(str, 52, stdin);
        str[strlen(str)-1] = 0;
        if(IsValidPS(str))
            printf("YES\n");
        else
            printf("NO\n");
    }
    return 0;
}

bool IsValidPS(const char* str) {
    stack<char> stk;
    for(int i=0; str[i] != '\0'; i++) {
        if(str[i]=='(' || str[i]=='[')
            stk.push(str[i]);
        if(str[i]==')') {
            if(stk.empty() || stk.top()!='(')
                return false;
            stk.pop();
        }
        if(str[i]==']') {
            if(stk.empty() || stk.top()!='[')
                return false;
            stk.pop();
        }
    }
    return stk.empty();
}